master/internal/task/allgather/allgather.go

```diff
@@ -0,0 +1,155 @@
+package allgather
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+	"github.com/determined-ai/determined/master/pkg/syncx/waitgroupx"
+)
+
+// ErrAllGatherTimeoutExceeded indicates that we not halt within the expected deadline.
+var ErrAllGatherTimeoutExceeded = fmt.Errorf(
+	"some ranks are taking a long time to connect to master" +
+		"during all gather; when running on kubernetes this may happen " +
+		"because only some of the pods have been scheduled; it is possible " +
+		"that some pods will never be scheduled without adding compute " +
+		"resources or pausing / killing other experiments in the cluster")
+
+// DefaultTimeout is the default timeout for all gather.
+var DefaultTimeout = 10 * time.Minute
+
+// Watcher contains a channel which can be polled for all gather completion.
+type Watcher struct {
+	C <-chan Result
+}
+
+// Result contains the information from a completed all gather.
+type Result struct {
+	Data []*structpb.Struct
+	Err  error
+}
+
+// Messages for all gathering.
+
+// WatchRequest begins or joins an all gather.
+type WatchRequest struct {
+	WatcherID uuid.UUID
+	NumPeers  int
+	Data      *structpb.Struct
+}
+
+// UnwatchRequest indicates the peer has disconnected.
+type UnwatchRequest struct {
+	WatcherID uuid.UUID
+}
+
+type AllGather struct {
+	watchers    map[uuid.UUID]chan Result
+	data        []*structpb.Struct
+	numPeers    *int
+	alreadyDone bool
+
+	readyCallback   func()
+	timeoutCallback func(error)
+
+	mu sync.Mutex
+	wg waitgroupx.Group
+}
+
+func New(readyCallback func(), timeoutCallback func(error)) *AllGather {
+	a := &AllGather{
+		readyCallback:   readyCallback,
+		timeoutCallback: timeoutCallback,
+		watchers:        map[uuid.UUID]chan Result{},
+		wg:              waitgroupx.WithContext(context.Background()),
+	}
+
+	a.wg.Go(a.run)
+
+	return a
+}
+
+func (g *AllGather) Join(msg WatchRequest) Watcher {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if _, ok := g.watchers[msg.WatcherID]; ok {
+		// If this peer has already connected, just respond with the watcher again. This is only
+		// possible if it disconnects and reconnects since the original actor ask blocks forever.
+		return Watcher{C: g.watchers[msg.WatcherID]}
+	}
+
+	// Channel is size 1 since data info will only ever be sent once and we'd rather not block.
+	w := make(chan Result, 1)
```

Comment:
praise: setting the channel size to 1 is a really clean way to minimize resource usage!